### PR TITLE
Workflow and job docs

### PR DESCRIPTION
**Pull request by beczkowb**

### CODE REVIEW

File: bigflow/workflow.py [hunk 1]
```diff
@@ -1,33 +1,40 @@
 from collections import OrderedDict
+from typing import Optional
+from .utils import now
 
 DEFAULT_SCHEDULE_INTERVAL = '@daily'
-
+# modyfikacja dagbuildera
```
Comment (iwko) [inline bigflow/workflow.py]: ?

File: docs/workflow-and-job.md [hunk 2]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
-You can also run the workflow and job within the module:
-
-```python
-hello_world_workflow.run()
-hello_world_workflow.run_job('hello_world')
-```
+## Job
 
-Finally, you can build and deploy created workflow to Cloud Composer:
+A job is just an object that has a unique `id` and implements the `run` method.
```
Comment (iwko) [inline docs/workflow-and-job.md:46]: Maybe we should use some kind of interface?

File: bigflow/workflow.py (hunk 1 above)
Comment (beczkowb) [reply bigflow/workflow.py]: deleted

File: docs/workflow-and-job.md [hunk 3]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
 
-```shell script
-bgf build;bgf deploy
-```
+The `id` parameter is a string that should be a valid Python variable name. For example - `'my_example_job'`, `'MY_EXAMPLE_JOB'`, `'job1234'`.
 
-### Job
+The run method takes a single argument - `runtime`. The `runtime` parameter is a date-time in the form of a string.
```
Comment (iwko) [inline docs/workflow-and-job.md:50]: We should put there example of this string

File: docs/workflow-and-job.md (hunk 2 above)
Comment (beczkowb) [reply docs/workflow-and-job.md:46]: I'm no fan of making java from python. There is an abstract class concept in python, but for me, it makes sense only if you have some logic on an abstract class.

File: docs/workflow-and-job.md [hunk 4]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
     def run(self, runtime):
-        print(f'Hello world at {runtime}!')
+        print(f'Running a simple job')
 
-hello_world_workflow = Workflow(workflow_id='hello_world_workflow', definition=[HelloWorldJob()])
+simple_workflow = Workflow(workflow_id='simple_workflow', definition=[SimpleJob()])
 ```
 
-Now, you can run the `hello_world_workflow`:
+You can run the workflow and job within a module (but we recommend using the BiggerQuery CLI):
```
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:29]: if we we recommend using BiggerQuery CLI, why it's not shown here?

File: docs/workflow-and-job.md (hunk 2 above)
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:46]: A job is just an object that has a unique `id` -> 
A job is just an object with a unique `id` and the `run` method.

File: docs/workflow-and-job.md [hunk 5]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
-```
+## Job
 
-Finally, you can build and deploy created workflow to Cloud Composer:
+A job is just an object that has a unique `id` and implements the `run` method.
 
-```shell script
-bgf build;bgf deploy
-```
+The `id` parameter is a string that should be a valid Python variable name. For example - `'my_example_job'`, `'MY_EXAMPLE_JOB'`, `'job1234'`.
```
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:48]: For example -
->
For example:

File: docs/workflow-and-job.md (hunk 3 above)
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:50]: &mdash; not minus

Comment (bartoszwalacik) [inline docs/workflow-and-job.md:50]: The Job `run` method

Comment (bartoszwalacik) [inline docs/workflow-and-job.md:50]: is a date-time value formatted as string?

File: docs/workflow-and-job.md [hunk 6]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
 
-### Job
+The run method takes a single argument - `runtime`. The `runtime` parameter is a date-time in the form of a string. 
+You can find more information about `runtime` and scheduling [in the chapter below](#workflow-scheduling-options).
```
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:51]: [in the chapter below](#workflow-scheduling-options)
->
in [workflow scheduling options](#workflow-scheduling-options)

File: docs/workflow-and-job.md [hunk 7]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
 
-Job is just an object that has a unique `id` and implements the `run` method.
+There are 2 additional parameters, that a job can supply - `retry_count` and `retry_pause_sec`. The `retry_count` parameter
```
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:53]: supply:

Comment (bartoszwalacik) [inline docs/workflow-and-job.md:53]: supply is unclear here, supply to whom?

File: docs/workflow-and-job.md [hunk 8]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
+determines how many times a job will be retried (in case of a failure). The `retry_pause_sec` says how long the pause between retries should be.
```
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:54]: `retry_pause_sec` or The `retry_pause_sec` parameter

File: docs/workflow-and-job.md [hunk 9]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
-starting from 2020-01-01. Then, the `run` method will be executed for each job with `runtime` 
-equals `"2020-01-01 07:00:00"`, `"2020-01-02 07:00:00"`, `"2020-01-03 07:00:00"` and so on.
-
-The `runtime` parameters is a string. It will be formatted as either `YYYY-MM-DD` or `YYYY-MM-DD hh-mm-ss`.
-It depends on the `Workflow` setup. By default, it will be date only - `YYYY-MM-DD`.
+## Workflow
 
-When you run a single job through the workflow class, without providing the `runtime` parameter, the `Workflow` class
-will pass the current date time as default.
+The `Workflow` class takes 2 main parameters - `workflow_id` and `definition`.
```
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:70]: parameters -
->
parameters:
or
parameters &mdash;

File: docs/workflow-and-job.md [hunk 10]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
-# Let's assume that now == '2020-01-02 01:11:00'
-simple_workflow = Workflow(workflow_id='simple_workflow', definition=[SimpleJob('1'), SimpleJob('2')])
-hello_world_workflow.run_job('2')
->>> '2020-01-02 01:11:00'
-```
+The `workflow_id` parameter is a string that should be a valid Python variable name. For example - `'my_example_workflow'`, `'MY_EXAMPLE_WORKFLOW'`, `'workflow1234'`.
 
-There are 2 additional parameters, that job can supply - `retry_count` and `retry_pause_sec`. The `retry_count` parameter
-determines how many times a job will be retried. The `retry_pause_sec` sets how long the pause between retries should be.
+The `Workflow` class arranges jobs into a DAG, through the `definition` parameter. There are 2 ways of specifying job arrangement. First one is passing a list
```
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:74]: DAG should be linked, maybe to Airflow DAG docs?

Comment (bartoszwalacik) [inline docs/workflow-and-job.md:74]: 2 ways -> two ways

Comment (bartoszwalacik) [inline docs/workflow-and-job.md:74]: When your jobs are executed sequentially, simply pass them in a list

File: docs/workflow-and-job.md [hunk 11]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
-## Workflow
+Output:
+```text
+Running job 1 at 2020-01-01
+Running job 2 at 2020-01-01
+```
 
-The `Workflow` class arranges jobs into a DAG. There are 2 ways of specifying job arrangement. First one is passing a list
-of jobs to `Workflow`:
+The second one is passing the `Definition` object, which allows you to create a graph.
```
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:101]: When some of your jobs are executed concurrently, pass them using the `Definition` object. It allows you to create a graph of jobs (DAG).

File: docs/workflow-and-job.md [hunk 12]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
 
+Let us say that we want to create the following DAG:
+    
+```
+       |--job2--|
+job1-->          -->job4 
+       |--job3--|
+```
+
+The implementation looks like this:
```
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:111]: The implementation looks like this:
->
The implementation:

File: docs/workflow-and-job.md [hunk 13]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
+```
 
-The `Workflow` class arranges jobs into a DAG. There are 2 ways of specifying job arrangement. First one is passing a list
-of jobs to `Workflow`:
+The second one is passing the `Definition` object, which allows you to create a graph.
 
+Let us say that we want to create the following DAG:
+    
+```
+       |--job2--|
```
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:106]:
```
       /--job2--\
job1-->          -->job4 
       \--job3--/
```

File: docs/workflow-and-job.md [hunk 14]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
+    job3: (job4,)
+}))
+graph_workflow.run()
+```
 
-example_workflow = Workflow(definition=[Job('1'), Job('2')])
+Output:
+```text
+Running job 1 at 2020-01-01
+Running job 2 at 2020-01-01
```
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:128]: is this Output worth to be written here?

File: docs/workflow-and-job.md [hunk 15]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
+Running job 3 at 2020-01-01
+Running job 4 at 2020-01-01
 ```
 
-The second one is passing `Definition` object, it allows you to create a graph:
+The `Workflow` class provides `run` and `run_job` methods. When you run a single job through a workflow object,
```
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:133]: the `run` and `run_job` methods

Comment (bartoszwalacik) [inline docs/workflow-and-job.md:133]: run a single job through ... it's unclear
what does it mean to run a job through a workflow object ?

File: docs/workflow-and-job.md [hunk 16]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
+```text
+Running job 1 at 2020-01-01
+Running job 2 at 2020-01-01
+Running job 3 at 2020-01-01
+Running job 4 at 2020-01-01
 ```
 
-The second one is passing `Definition` object, it allows you to create a graph:
+The `Workflow` class provides `run` and `run_job` methods. When you run a single job through a workflow object, 
+without providing the `runtime` parameter, the `Workflow` class will pass the current date-time (local time) as default.
```
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:134]: will pass -> passes

File: docs/workflow-and-job.md [hunk 17]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
-from biggerquery.build import Workflow
-from biggerquery.build import Definition
-from biggerquery.build import WorkflowJob
+simple_workflow = Workflow(
+    workflow_id='simple_workflow',
+    runtime_as_datetime=True,
+    definition=[Job('1')])
+simple_workflow.run_job('1')
+simple_workflow.run()
+simple_workflow.run_job('1', '1970-01-01')
```
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:144]: why do we care about this case? running jobs in a workflow with different runtime?

File: docs/workflow-and-job.md [hunk 18]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
+Output:
+```text
+Running job 1 at 2020-01-02 01:11:00
+Running job 1 at 2020-01-02 01:11:00
+Running job 1 at 1970-01-01
+Running job 1 at 1970-01-01
+```
 
-job1, job2, job3, job4 = Job('1'), Job('2'), Job('3'), Job('4')
+The `Workflow.run` method ignores job parameters like `retry_count` and `retry_pause_sec`. It executes the workflow in a sequential (non-parallel) way.
```
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:156]: It executes the workflow ->  It executes a workflow

File: docs/workflow-and-job.md [hunk 19]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
-Definition({
-    create_quality_metric_table_job: (calculate_box_quality_metric_job, calculate_offer_quality_metric_job)
-})
+### The `runtime` parameter
+
+The most important parameter for a workflow is `runtime`. BigFlow workflows process data in batches, 
+where batch means: all units of data having timestamps within a given period. The `runtime` parameter defines this period.
+
+When a workflow is deployed on Airflow, the `runtime` parameter is taken from Airflow `execution_date`. 
+It will be formatted as either `YYYY-MM-DD` or `YYYY-MM-DD hh-mm-ss`. It depends on the `Workflow` setup.
```
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:167]: 😄

File: docs/workflow-and-job.md [hunk 20]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
...
+
+The `Workflow` class has some additional parameters.
```
Comment (bartoszwalacik) [inline docs/workflow-and-job.md:169]: without some

has two additional parameters:

File: docs/workflow-and-job.md [hunk 21]
```diff
@@ -1,6 +1,6 @@
-## Workflows and jobs
+# Workflow & job
 
-### Overview
+## Overview
```
[the remark this answers is not in the document]
Comment (beczkowb) [reply docs/workflow-and-job.md:3]: done

[the remark this answers is not in the document]
Comment (beczkowb) [reply docs/workflow-and-job.md:3]: done

[the remark this answers is not in the document]
Comment (beczkowb) [reply docs/workflow-and-job.md:3]: done

File: docs/workflow-and-job.md [hunk 22]
```diff
@@ -10,151 +10,222 @@ It can execute anything that can be executed from the Python code, for example:
 * BigQuery query
 * Any Python code.
 
-The simplest workflow you can create, looks like this:
+The simplest workflow you can create looks like this:
 
+[`./docs_examples/workflow_and_job/simple_workflow_and_job.py`](./docs_examples/workflow_and_job/simple_workflow_and_job.py)
```
[the remark this answers is not in the document]
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

File: docs/workflow-and-job.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

[the remark this answers is not in the document]
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

[the remark this answers is not in the document]
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

Comment (beczkowb) [reply docs/workflow-and-job.md]: done

File: docs/workflow-and-job.md (hunk 2 above)
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

File: docs/workflow-and-job.md (hunk 5 above)
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

File: docs/workflow-and-job.md (hunk 3 above)
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

Comment (beczkowb) [reply docs/workflow-and-job.md]: done

Comment (beczkowb) [reply docs/workflow-and-job.md]: done

File: docs/workflow-and-job.md (hunk 6 above)
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

File: docs/workflow-and-job.md (hunk 7 above)
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

Comment (beczkowb) [reply docs/workflow-and-job.md]: done

File: docs/workflow-and-job.md (hunk 8 above)
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

File: docs/workflow-and-job.md (hunk 9 above)
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

File: docs/workflow-and-job.md (hunk 10 above)
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

Comment (beczkowb) [reply docs/workflow-and-job.md]: done

Comment (beczkowb) [reply docs/workflow-and-job.md]: done

File: docs/workflow-and-job.md (hunk 11 above)
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

File: docs/workflow-and-job.md (hunk 12 above)
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

File: docs/workflow-and-job.md (hunk 13 above)
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

File: docs/workflow-and-job.md (hunk 14 above)
Comment (beczkowb) [reply docs/workflow-and-job.md:130]: done

File: docs/workflow-and-job.md (hunk 15 above)
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

Comment (beczkowb) [reply docs/workflow-and-job.md]: done

File: docs/workflow-and-job.md (hunk 16 above)
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

File: docs/workflow-and-job.md (hunk 17 above)
Comment (beczkowb) [reply docs/workflow-and-job.md:146]: done

File: docs/workflow-and-job.md (hunk 18 above)
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

File: docs/workflow-and-job.md (hunk 19 above)
Comment (beczkowb) [reply docs/workflow-and-job.md:161]: done

File: docs/workflow-and-job.md (hunk 20 above)
Comment (beczkowb) [reply docs/workflow-and-job.md]: done

**Pull request merged**